Conversion from Gdoc to Markdown (#96)
ippontech/blog-usa: posts/a-guinea-pig-in-the-cloud.md

---
authors:
- Bertrand PINEL
tags:
- DevOps
- Cloud
- Serverless
- Architectures
- AWS
- Ember
- Front
date: 2019-02-27T17:27:07.000Z
title: "A Guinea Pig in the Cloud"
image: https://github.com/bpinel/blog-usa/blob/master/images/2019/02/guinea-pigs.png
---

# Introduction

This article is not about the latest performance of the Chinese space industry and it will not hurt any pets. It is just a teaser about an idea that popped in my mind a few months ago and is the basis for a small tool I have been writing since that moment. 

The idea is very simple. On one hand, thanks to the Cloud Providers, it is now very easy to host both a web site and some REST services and to benefit from managed services like Databases. On the other hand, JavaScript frameworks like [Angular](https://angular.io/), [React](https://reactjs.org/) or [Ember](https://emberjs.com/)... 

> ![Ember logo](https://www.emberjs.com/images/brand/ember_Tomster-Lockup-3a0f7b78.png)

Reviewer: Looks like this link is failing. You could try this:
https://emberjs.com/images/tomster-twitter-card-c38bcf1af22002906470c99fed09c044.png

![Ember logo](https://emberjs.com/images/tomster-twitter-card-c38bcf1af22002906470c99fed09c044.png)

Reviewer: Other than that, the article looks great! I'm going to try this out when I get the chance

Author: Thanks for your comment and suggestion. Hope you will find time to try my little tool.
I have updated the image logo according to your suggestion !
Thanks again.

> ![Ember logo](https://www.emberjs.com/images/brand/ember_Tomster-Lockup-3a0f7b78.png)

Author: ```suggestion
![Ember logo](https://emberjs.com/images/tomster-twitter-card-c38bcf1af22002906470c99fed09c044.png)
```

This solution should be available in three modes : 

*   The **Developer mode** needs no Internet connection and looks like the standard way of building an [Ember CLI](https://ember-cli.com/) application. However it adds a new key feature: the possibility to import a [JDL](https://start.jhipster.tech/jdl-studio/) file describing the application object model, directly creating both the different entities and their smart mocked implementations through the use of [Ember CLI Mirage](https://www.ember-cli-mirage.com/) and [Faker.js](https://github.com/marak/Faker.js/). Quoting Ember CLI Mirage documentation, this mode allows the developer to _share a functional prototype of his app that runs entirely in the client - before writing a single line of your API_. The following diagram illustrate the way everything works together using the different concepts proposed by Ember:

> ![Developper mode](https://github.com/bpinel/blog-usa/master/images/2019/02/ember-aws-ehipster-dev.png)

Author: ```suggestion
![Developer mode](https://github.com/bpinel/blog-usa/blob/master/images/2019/02/ember-aws-ehipster-dev.png)
```

*   The **Integration mode** uses the proxy feature of Ember to rely on a Cloud back-end persisting data in [DynamoDB](https://aws.amazon.com/dynamodb/). To do that, it benefits from the fact that Ember Data heavily relies on the JSON API specification. The cloud part of the framework therefore has the goal of setting up this Cloud infrastructure, thanks to the [API Gateway](https://aws.amazon.com/api-gateway/) delivering REST services with the right semantics and transferring the payload of the message to a lambda function interacting with a Dynamo database. This mode is great for thoroughly testing the application before complete deployment to the Cloud. The following diagram illustrate this behaviour and the way Mirage has been substituted to the real back-end:

> ![Integration mode](https://github.com/bpinel/blog-usa/master/images/2019/02/ember-aws-ehipster-int.png)

Author: ```suggestion
![Integration mode](https://github.com/bpinel/blog-usa/blob/master/images/2019/02/ember-aws-ehipster-int.png)
```

*   The **production mode** just pushes the web resources to a S3 bucket exposed through CloudFront. The application can then be accessed all over the world!

>  ![Production mode](https://github.com/bpinel/blog-usa/master/images/2019/02/ember-aws-ehipster-prod.png)

Author: ```suggestion
 ![Production mode](https://github.com/bpinel/blog-usa/blob/master/images/2019/02/ember-aws-ehipster-prod.png)
```

```
entity User {
 login String required minlength(4),
 passwordHash String required minlength(8),
 firstName String required minlength(3) maxlength(50),
 lastName String,
 email String,
 imageUrl String,
 activated Boolean,
 createdBy String,
 createdDate LocalDate,
 lastModifiedBy String,
 lastModifiedDate Instant 
}

entity Authority {
 name String required maxlength(30)
}

entity Todo {
  title String required,
  completed Boolean
}

relationship ManyToMany {
	User{user(login)} to Authority{authority(name)}
}

relationship OneToMany {
 	User{user(login)} to Todo{todo}
}
```

Note that **Authority** and **User** objects are just taken from the JHipster model and can be represented by the following diagram:

> ![JDL model](https://github.com/bpinel/blog-usa/blob/master/images/2019/02/jdl-todo.png)

Reviewer 2: ✅ FYI, this image, and all the subsequent images, work correctly.

Author: I miss the /blob part in the previous URL... Including in the one referenced from the metadata (tag 'image:'). Can you correct it ?

Just store the previous object model into a file named **<code>todo.jh</code></strong> inside the <em>guineapig_inthecloud</em> directory and play the following command: 

```shell
ember g jdl-importer todo.jh
```

This blueprint will simply generate all the object available in the JDL model, the factories for generating sample mock data and the pages to manage them. So starting the server with:

```
ember s
```

> 

Reviewer 2: I've been following along with the example, but when I reach this point, the startup process dumps errors.

```
>> ember s
Could not start watchman
Visit https://ember-cli.com/user-guide/#watchman for more info.

guineapig/templates/entity-factory.hbs
  1:23  error  Incorrect indentation for `
    Available entities :
    ` beginning at L1:C23. Expected `
    Available entities :
    ` to be at an indentation of 2 but was found at 4.  block-indentation
  3:4  error  Incorrect indentation for `<select>` beginning at L3:C4. Expected `<select>` to be at an indentation of 2 but was found at 4.  block-indentation
  4:8  error  Incorrect indentation for `{{#each}}` beginning at L4:C8. Expected `{{#each}}` to be at an indentation of 6 but was found at 8.  block-indentation
  5:12  error  Incorrect indentation for `<option>` beginning at L5:C12. Expected `<option>` to be at an indentation of 10 but was found at 12.  block-indentation

guineapig/templates/entity-factory/authority.hbs
  4:0  error  Incorrect indentation for `<h3>` beginning at L4:C0. Expected `<h3>` to be at an indentation of 2 but was found at 0.  block-indentation
  6:0  error  Incorrect indentation for `{{#models-table}}` beginning at L6:C0. Expected `{{#models-table}}` to be at an indentation of 2 but was found at 0.  block-indentation
  12:0  error  Incorrect indentation for `{{#bs-button}}` beginning at L12:C0. Expected `{{#bs-button}}` to be at an indentation of 2 but was found at 0.  block-indentation
  16:0  error  Incorrect indentation for `{{#bs-modal}}` beginning at L16:C0. Expected `{{#bs-modal}}` to be at an indentation of 2 but was found at 0.  block-indentation
  29:20  error  Incorrect indentation for `validated-form` beginning at L23:C4. Expected `{{/validated-form}}` ending at L29:C20 to be at an indentation of 4 but was found at 1.  block-indentation
  24:3  error  Incorrect indentation for `{{f.input}}` beginning at L24:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  25:3  error  Incorrect indentation for `{{f.input}}` beginning at L25:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  28:2  error  Incorrect indentation for `{{f.submit}}` beginning at L28:C2. Expected `{{f.submit}}` to be at an indentation of 6 but was found at 2.  block-indentation

guineapig/templates/entity-factory/todo.hbs
  4:0  error  Incorrect indentation for `<h3>` beginning at L4:C0. Expected `<h3>` to be at an indentation of 2 but was found at 0.  block-indentation
  6:0  error  Incorrect indentation for `{{#models-table}}` beginning at L6:C0. Expected `{{#models-table}}` to be at an indentation of 2 but was found at 0.  block-indentation
  12:0  error  Incorrect indentation for `{{#bs-button}}` beginning at L12:C0. Expected `{{#bs-button}}` to be at an indentation of 2 but was found at 0.  block-indentation
  16:0  error  Incorrect indentation for `{{#bs-modal}}` beginning at L16:C0. Expected `{{#bs-modal}}` to be at an indentation of 2 but was found at 0.  block-indentation
  30:20  error  Incorrect indentation for `validated-form` beginning at L23:C4. Expected `{{/validated-form}}` ending at L30:C20 to be at an indentation of 4 but was found at 1.  block-indentation
  24:3  error  Incorrect indentation for `{{f.input}}` beginning at L24:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  25:3  error  Incorrect indentation for `{{f.input}}` beginning at L25:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  26:3  error  Incorrect indentation for `{{f.input}}` beginning at L26:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  29:2  error  Incorrect indentation for `{{f.submit}}` beginning at L29:C2. Expected `{{f.submit}}` to be at an indentation of 6 but was found at 2.  block-indentation

guineapig/templates/entity-factory/user.hbs
  4:0  error  Incorrect indentation for `<h3>` beginning at L4:C0. Expected `<h3>` to be at an indentation of 2 but was found at 0.  block-indentation
  6:0  error  Incorrect indentation for `{{#models-table}}` beginning at L6:C0. Expected `{{#models-table}}` to be at an indentation of 2 but was found at 0.  block-indentation
  12:0  error  Incorrect indentation for `{{#bs-button}}` beginning at L12:C0. Expected `{{#bs-button}}` to be at an indentation of 2 but was found at 0.  block-indentation
  16:0  error  Incorrect indentation for `{{#bs-modal}}` beginning at L16:C0. Expected `{{#bs-modal}}` to be at an indentation of 2 but was found at 0.  block-indentation
  46:20  error  Incorrect indentation for `validated-form` beginning at L23:C4. Expected `{{/validated-form}}` ending at L46:C20 to be at an indentation of 4 but was found at 1.  block-indentation
  24:3  error  Incorrect indentation for `{{f.input}}` beginning at L24:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  25:3  error  Incorrect indentation for `{{f.input}}` beginning at L25:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  26:3  error  Incorrect indentation for `{{f.input}}` beginning at L26:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  27:3  error  Incorrect indentation for `{{f.input}}` beginning at L27:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  28:3  error  Incorrect indentation for `{{f.input}}` beginning at L28:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  29:3  error  Incorrect indentation for `{{f.input}}` beginning at L29:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  30:3  error  Incorrect indentation for `{{f.input}}` beginning at L30:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  31:3  error  Incorrect indentation for `{{f.input}}` beginning at L31:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  32:3  error  Incorrect indentation for `{{#f.input}}` beginning at L32:C3. Expected `{{#f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  35:3  error  Incorrect indentation for `{{f.input}}` beginning at L35:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  36:3  error  Incorrect indentation for `{{#f.input}}` beginning at L36:C3. Expected `{{#f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  39:3  error  Incorrect indentation for `{{f.input}}` beginning at L39:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  42:3  error  Incorrect indentation for `{{f.input}}` beginning at L42:C3. Expected `{{f.input}}` to be at an indentation of 6 but was found at 3.  block-indentation
  45:2  error  Incorrect indentation for `{{f.submit}}` beginning at L45:C2. Expected `{{f.submit}}` to be at an indentation of 6 but was found at 2.  block-indentation
  33:4  error  Incorrect indentation for `{{pikaday-input}}` beginning at L33:C4. Expected `{{pikaday-input}}` to be at an indentation of 5 but was found at 4.  block-indentation
  37:4  error  Incorrect indentation for `{{pikaday-input}}` beginning at L37:C4. Expected `{{pikaday-input}}` to be at an indentation of 5 but was found at 4.  block-indentation

===== 4 Template Linting Errors

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/adapters/authority.js
  5:5  error  Only string, number, symbol, boolean, null, undefined, and function are allowed as default properties  ember/avoid-leaking-state-in-ember-objects

✖ 1 problem (1 error, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/adapters/todo.js
  5:5  error  Only string, number, symbol, boolean, null, undefined, and function are allowed as default properties  ember/avoid-leaking-state-in-ember-objects

✖ 1 problem (1 error, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/adapters/user.js
  5:5  error  Only string, number, symbol, boolean, null, undefined, and function are allowed as default properties  ember/avoid-leaking-state-in-ember-objects

✖ 1 problem (1 error, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/controllers/entity-factory.js
  4:5  error  Only string, number, symbol, boolean, null, undefined, and function are allowed as default properties  ember/avoid-leaking-state-in-ember-objects

✖ 1 problem (1 error, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/controllers/entity-factory/authority.js
  37:7  error  Unexpected console statement  no-console
  40:7  error  Unexpected console statement  no-console
  48:5  error  Unexpected console statement  no-console

✖ 3 problems (3 errors, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/controllers/entity-factory/todo.js
  42:7  error  Unexpected console statement  no-console
  45:7  error  Unexpected console statement  no-console
  53:5  error  Unexpected console statement  no-console

✖ 3 problems (3 errors, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/controllers/entity-factory/user.js
   91:7  error  Unexpected console statement  no-console
   94:7  error  Unexpected console statement  no-console
  102:5  error  Unexpected console statement  no-console

✖ 3 problems (3 errors, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/routes/entity-factory/authority.js
  5:12  error  Use import { hash } from 'rsvp'; instead of using Ember.RSVP.hash  ember/new-module-imports
  5:12  error  'Ember' is not defined                                             no-undef
  6:2   error  Mixed spaces and tabs                                              no-mixed-spaces-and-tabs

✖ 3 problems (3 errors, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/routes/entity-factory/todo.js
  5:12  error  Use import { hash } from 'rsvp'; instead of using Ember.RSVP.hash  ember/new-module-imports
  5:12  error  'Ember' is not defined                                             no-undef
  6:2   error  Mixed spaces and tabs                                              no-mixed-spaces-and-tabs

✖ 3 problems (3 errors, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/routes/entity-factory/user.js
  5:12  error  Use import { hash } from 'rsvp'; instead of using Ember.RSVP.hash  ember/new-module-imports
  5:12  error  'Ember' is not defined                                             no-undef
  6:2   error  Mixed spaces and tabs                                              no-mixed-spaces-and-tabs

✖ 3 problems (3 errors, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/validations/authority.js
  1:27  error  'validateLength' is defined but never used  no-unused-vars
  1:43  error  'validateFormat' is defined but never used  no-unused-vars

✖ 2 problems (2 errors, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/validations/todo.js
  1:27  error  'validateLength' is defined but never used  no-unused-vars
  1:43  error  'validateFormat' is defined but never used  no-unused-vars

✖ 2 problems (2 errors, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/app/validations/user.js
  1:43  error  'validateFormat' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/mirage/mirage/models/authority.js
  1:17  error  'belongsTo' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/mirage/mirage/models/todo.js
  1:17  error  'belongsTo' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)

/Users/pasha/dev/ippon/guineapig-demo/guineapig/mirage/mirage/models/user.js
  1:17  error  'belongsTo' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)

Build successful (30270ms) – Serving on http://localhost:4200/

Slowest Nodes (totalTime => 5% )              | Total (avg)         
----------------------------------------------+---------------------
Babel: ember-data (2)                         | 4308ms (2154 ms)    
Babel: guineapig (2)                          | 2846ms (1423 ms)    
Babel: Modules for Templates (7)              | 2152ms (307 ms)     
Babel: ember-test-helpers (2)                 | 1705ms (852 ms)     
Babel: ember-models-table (1)                 | 1665ms              
Bundler (1)                                   | 1583ms              
Babel: ember-bootstrap (1)                    | 1572ms              

ls
cleaning up...
```

This unfortunately means that I can't continue verifying the _content_ of the blog post; further comments will be limited to simple proofreading.

Author: These errors are for the most part simply lint errors. None of them prevents the software from running. 
As stated in the documentation of my npm addon, they can be turned off by adding to the file .template-lintrc.js the following rules:
  rules: {
    "attribute-indentation": false,
    "block-indentation": false
  }

Author: I have just pushed to npm a new version of the plugin which integrate the added lines to configure lint checks. 
The number of "errors" has decreased but you still have warning about:
- Using console.log (Unexpected console statement)
- Mixed spaces and tabs
- Variables defined but never used
These are definitely warnings and not errors and doesn't prevent the application from running.

A JSON API compliant back-end should implement the HTTP vocabulary as depicted below :

![JSON API](https://github.com/bpinel/blog-usa/blob/master/images/2019/02/API-Gateway.png)

To achieve the implementation of this backend, we will rely mainly on three AWS services : 

*   A new API declared in the **API Gateway** that support the awaited REST semantic,
*   A Lambda function written on Javascript and running on Node.js 8.1, systematically called by the API and in charge of the logical processing of incoming parameters and also of the handling of potential relationships
*   A DynamoDB table, in charge of persisting all the processed objects independently of their content.

This main infrastructure for the backend is completed by the following elements:

*   Three **S3 buckets**:
    *   The first one (_lambda-jsonapi-code-bucket_) is simply used for provisioning the lambda code at its creation time,

>     *   The other two (_ember-aws-ehipster-staging_ et _ember-aws-ehipster-production_) will be used later to store the 'staging' and 'production' versions of the application.  

Reviewer 2: ```suggestion
    *   The other two (_ember-aws-ehipster-staging_ and _ember-aws-ehipster-production_) will be used later to store the 'staging' and 'production' versions of the application.  
```
🇺🇸 😆

*   An **IAM role** (_lambda_jsonapi_) allowing the lambda function to get access to the DynamoDB database,
*   Two entries in **CloudFront** corresponding respectively to the buckets of 'staging' and 'production'

### Building the backend infrastructure in AWS

> Rather than going through a manual construction of this infrastructure thanks to the AWS Web console, I have privilege an automated approach based on [terraform](https://www.terraform.io/) scripts, largely easier to read than their [CloudFormation](https://aws.amazon.com/fr/cloudformation/) counterparts. All necessary files are available in the cloud directory of the gitlab project: [https://gitlab.ippon.fr/bpinel/ember-aws-ehipster/tree/master/cloud](https://gitlab.ippon.fr/bpinel/ember-aws-ehipster/tree/master/cloud) that you should get on your local computer.

Reviewer 2: This first sentence is phrased awkwardly.

Author: Tell me if the following sentences are clearer. 
```suggestion
Rather than going through a step-by-step construction of this infrastructure thanks to the AWS Web console, I have privilege an automated approach based on [terraform](https://www.terraform.io/) scripts. To my mind, Terraform scripts are definitely easier to write and to maintain than their [CloudFormation](https://aws.amazon.com/fr/cloudformation/) counterparts. All necessary files are available in the cloud directory of the gitlab project: [https://gitlab.ippon.fr/bpinel/ember-aws-ehipster/tree/master/cloud](https://gitlab.ippon.fr/bpinel/ember-aws-ehipster/tree/master/cloud) that you should download on your local computer.
```

Clicking on this API and going to its Staging, you should find a 'staging' deployment. You should grab here its Invoke URL:

![API Gateway web console](https://github.com/bpinel/blog-usa/blob/master/images/2019/02/api-gateway-console-02.png)

As a matter of fact, you need to tell your Ember application to use this URL as the Gateway for all calls done though Ember Data. Everything happens in the file_ config/environment.js_ in the self explanatory lines:

```javascript
 if (environment === 'production') {
    // here you can enable a production-specific feature
    ENV.gatewayURL = 'https://xxxxxxxxxxx.execute-api.us-east-1.amazonaws.com/staging';
    //ENV.gatewayURL = process.env.STAGING_GATEWAY_URL;
    console.log("Setting staging gatewayURL to "+ENV.gatewayURL);
  }
  if (environment === 'staging') {
    // here you can enable a staging-specific feature
    ENV.gatewayURL = 'https://xxxxxxxxxxx.execute-api.us-east-1.amazonaws.com/staging';
    //ENV.gatewayURL = process.env.PRODUCTION_GATEWAY_URL;
    console.log("Setting production gatewayURL to "+ENV.gatewayURL);
  }
```

### A little test before going further

> To check the good provisioning of the cloud infrastructure, you can easily launch a test. To do that, just go to the API Gateway Web console, look for the newly created _JsonApiRestGateway_ API and especially to the GET méthod for /{type} and click on the 'Test' button:

Reviewer 2: This is also awkwardly phrased... and has a sneaky French survivor toward the end. 😄 

```suggestion
To verify that the cloud infrastructure is properly provisioned, you can open the API Gateway Web console, look for the newly created `JsonApiRestGateway` API –– specifically, the `GET` method for `/{type}` –– and click on the **Test** button:
```

In our case, we will define 3 .env files (one global, one for the staging environment and the latest for the production environment) like the following ones :

```
# .env
AWS_KEY='XXXXXXXXXXXXXXXXXXXX'
AWS_SECRET='XxXxxxxXXXXXX/XxxxXXXxx/xxxXXX'
AWS_REGION='us-east-1'
```

For the other two files, you have to collect in the CloudFront Web console, the distribution identifier for each target environment and paste them in the corresponding .env file:

![CloudFront access](https://github.com/bpinel/blog-usa/blob/master/images/2019/02/list-authorities.png)

(Please note that this console also give you the URL exposed at CloudFront level that you will need later on) 

```
# .env.staging
STAGING_REGION='us-east-1'
STAGING_BUCKET='ember-aws-ehipster-staging'
STAGING_DISTRIBUTION='E10C67WZNAPJ4B'
```

```
# .env.production
PRODUCTION_REGION='us-east-1'
PRODUCTION_BUCKET='ember-aws-ehipster-production'
PRODUCTION_DISTRIBUTION='E1VOWVNE5HYGGC'
```

The final deployment on the AWS S3 buckets is them simply done through the following command (here for the staging environment):

```shell
ember deploy staging --verbose
```

> The --verbose can be skipped but show you the different processing on the files, the optimizations done and the copy to S3 bucket.

Reviewer 2: Keep code examples/references in fixed-width font.
```suggestion
The `--verbose` can be skipped but show you the different processing on the files, the optimizations done and the copy to S3 bucket.
```
